### PR TITLE
Dockerfile

### PR DESCRIPTION
**Pull request by FrancescoSaverioZuppichini**
## Description
This PR adds a `Dockerfile` to quickly run the `finetune.py` script without losing sanity

## Usage

Build the container

```bash
docker build -t alpaca-lora .
```

Run it, 

**Training**:

```bash
docker run --gpus=all --ipc=host --shm-size 64g \
  -v ${HOME}/.cache:/root/.cache \
  -v ${PWD}/lora-alpaca:/workspace/lora-alpaca \
  --rm -it alpaca-lora finetune.py \
  --base_model 'decapoda-research/llama-7b-hf' \
  --data_path 'yahma/alpaca-cleaned' \
  --output_dir './lora-alpaca' \
  --batch_size 128 \
  --micro_batch_size 4 \
  --num_epochs 3 \
  --learning_rate 1e-4 \
  --cutoff_len 512 \
  --val_set_size 2000 \
  --lora_r 8 \
  --lora_alpha 16 \
  --lora_dropout 0.05 \
  --lora_target_modules '[q_proj,v_proj]' \
  --train_on_inputs \
  --group_by_length
```

**Generation**

```bash
docker run --gpus=all --ipc=host --shm-size 64g \
  -v ${HOME}/.cache:/root/.cache \
  -v ${PWD}/lora-alpaca:/workspace/lora-alpaca \
  -p 7860:7860 \
  --rm -it alpaca-lora generate.py
  ```

Then natigate to `https://localhost:7860`

all good!

image

### TODOs

- [x] add more env variables to control training

### CODE REVIEW

**Comment by dittops** [comment]
A dockerfile is really needed to clear the env-related issues. Thanks for initiating this. I have tried running this on my machine, but I'm getting this error after running docker.

image

**Comment by FrancescoSaverioZuppichini** [comment]
@dittops thanks for trying it out, I don't see the issue. Unfortunately, till tomorrow I am travelling. Just to be sure, are you using the code from the `main` branch?

**Comment by AngainorDev** [comment]
> I'm getting this error after running docker.

Would you have more than 1 gpu by chance?
Can you try exposing only one of them to Docker?

**Comment by thekevshow** [comment]
Good stuff, I was dockerizing this myself, but you beat me too finishing it. I will check this out and see if there are any issues.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    software-properties-common \
+    && add-apt-repository ppa:deadsnakes/ppa \
+    && apt install -y python3.10 \
+    && rm -rf /var/lib/apt/lists/*
+WORKDIR /workspace
+COPY requirements.txt requirements.txt
+RUN curl -sS https://bootstrap.pypa.io/get-pip.py | python3.10
+RUN python3.10 -m pip install -r requirements.txt
+# install torch==2.1
+RUN python3.10 -m pip install numpy --pre torch --force-reinstall --index-url https://download.pytorch.org/whl/nightly/cu118
```
Comment (rohitgr7) [inline Dockerfile:16]: you should combine all the RUN calls here. Reduces the number of layers and possibly the image size.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,18 @@
+FROM nvidia/cuda:11.8.0-devel-ubuntu22.04
```
Comment (rohitgr7) [inline Dockerfile:1]: maybe use pytorch image itself: `FROM pytorch/pytorch:1.13.1-cuda11.6-cudnn8-runtime`
that way you can few installation steps

Comment (FrancescoSaverioZuppichini) [reply Dockerfile:1]: We want to use pytorch 2.1, so we can compile the model

File: Dockerfile (hunk 1 above)
Comment (FrancescoSaverioZuppichini) [reply Dockerfile]: Oh right, thanks a lot :)

File: README.md [hunk 3]
```diff
@@ -28,6 +28,20 @@ pip install -r requirements.txt
 
 2. If bitsandbytes doesn't work, [install it from source.](https://github.com/TimDettmers/bitsandbytes/blob/main/compile_from_source.md) Windows users can follow [these instructions](https://github.com/tloen/alpaca-lora/issues/17).
 
+#### Docker
+
+You can use `docker` with build in support for `torch==2.1.0` and `bytesands`
```
[the remark this answers is not in the document]
Comment (FrancescoSaverioZuppichini) [reply README.md]: Thanks :)

File: Dockerfile (hunk 2 above)
Comment (rohitgr7) [reply Dockerfile:1]: I don't think that's released yet: https://pypi.org/project/torch/

if 2.0 works: `pytorch/pytorch:2.0.0-cuda11.7-cudnn8-runtime`

Comment (FrancescoSaverioZuppichini) [reply Dockerfile:1]: I've sticked with the current solution for now :)

**Comment by FrancescoSaverioZuppichini** [comment]
Changed `finetune.py` to use env variables, handy if you want to quickly change some finetuning parameters

Updated the doc accordingly

**Comment by FrancescoSaverioZuppichini** [comment]
The docker image now only exposes `python3.10`, thus we can also run the `generate` script by 

```
docker run --gpus=all --ipc=host --shm-size 64g \
  -v ${HOME}/.cache:/root/.cache \
  -v ${PWD}/lora-alpaca:/workspace/lora-alpaca \
  -p 7860:7860 \
  --rm -it alpaca-lora generate.py
  ```

**Comment by FrancescoSaverioZuppichini** [comment]
Resolved conflicts, remove env variables

**Comment by FrancescoSaverioZuppichini** [comment]
@tloen I think this is in a good state now, fell free to add any comment :)

**Comment by tloen** [comment]
I'll try to get this tested in the next day or two. Good work, and thanks for hanging in with all the interface changes!

File: README.md [hunk 4]
```diff
@@ -73,6 +106,29 @@ python finetune.py \
...
+  --learning_rate 1e-4 \
+  --cutoff_len 512 \
+  --val_set_size 2000 \
+  --lora_r 8 \
+  --lora_alpha 16 \
+  --lora_dropout 0.05 \
+  --lora_target_modules '[q_proj,v_proj]' \
+  --train_on_inputs \
+  --group_by_length
+
```
Comment (gramster) [inline README.md:131]: You are missing the closing quotes on the code block here.

File: README.md [hunk 5]
```diff
@@ -25,18 +25,51 @@ Without hyperparameter tuning, the LoRA model produces outputs comparable to the
...
+  -v ${HOME}/.cache:/root/.cache \
+  -v ${PWD}/lora-alpaca:/workspace/lora-alpaca \
+  --rm -it alpaca-lora training.py
+```
+
+You mapped the host `.cache` folder to the container to persist the original model checkpoint in future runs.
+
+### Inference (`generate.py`)
+
     ```bash
```
Comment (gramster) [inline README.md:52]: Not your error but this should have code block quotes or be indented, not both.

File: README.md [hunk 6]
```diff
@@ -25,18 +25,51 @@ Without hyperparameter tuning, the LoRA model produces outputs comparable to the
 
 1. Set environment variables, or modify the files referencing `BASE_MODEL`:
```
Comment (gramster) [inline README.md:27]: Looks like some stuff got clobbered here.

**Comment by tloen** [comment]
Merged in 4367a43fcb8c47ff1e520777ad7e60b622edea4d

**Comment by FrancescoSaverioZuppichini** [comment]
🥳

**Pull request closed without merging**